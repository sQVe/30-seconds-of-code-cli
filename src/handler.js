const snippets = require('./snippets');
const { any, sortBy, startsOrEndsWith } = require('./helpers');

const fuzzyMatch = query => x => new RegExp(query).test(x.id);
const getSnippet = id => snippets.find(x => x.id === id);
const getSnippetsByTag = id => snippets.filter(x => any(x.tags, y => y === id));

function setSearchRelevance(query) {
  return x => {
    let relevance = query.length * 10;

    if (x.id === query) return { ...x, relevance };
    if (startsOrEndsWith(x.id, query)) relevance *= 0.75;
    else relevance /= 2;

    return {
      ...x,
      relevance: relevance / (x.id.replace(query, '').length + 1),
    };
  };
}

const searchSnippets = query =>
  snippets
    .filter(fuzzyMatch(query))
    .map(setSearchRelevance(query))
    .sort(sortBy('relevance'));

module.exports = {
  getSnippet,
  getSnippetsByTag,
  searchSnippets,
};
