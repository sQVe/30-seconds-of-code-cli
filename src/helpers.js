const any = (arr, fn = Boolean) => arr.some(fn);
const is = (type, val) => ![, null].includes(val) && val.constructor === type; // eslint-disable-line no-sparse-arrays
const last = arr => arr[arr.length - 1];
const mapValues = (obj, fn) =>
  Object.keys(obj).reduce((acc, k) => {
    acc[k] = fn(obj[k], k, obj);
    return acc;
  }, {});
const startsOrEndsWith = (val, query) =>
  val.startsWith(query) || val.endsWith(query);

const enforceSingleNewLine = x =>
  is(String, x) ? x.replace(/[\r\n]+/g, '\n').replace(/[\r\n]*$/, '') : x;

function sortBy(key) {
  return (a, b) => {
    if (a[key] === b[key]) return 0;
    return a[key] < b[key] ? 1 : -1;
  };
}

module.exports = {
  any,
  enforceSingleNewLine,
  is,
  last,
  mapValues,
  sortBy,
  startsOrEndsWith,
};
