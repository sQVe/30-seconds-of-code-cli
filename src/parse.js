const fs = require(`fs`);
const path = require('path');
const { promisify } = require('util');

const snippets = require('../submodules/30-seconds-of-code/snippet_data/snippets.json')
  .data;
const { isString, mapValues } = require('./helpers');

const FILE_NAME = 'snippets.json';

const enforceSingleNewLine = x =>
  isString(x) ? x.replace(/[\r\n]+/g, '\n').replace(/[\r\n]*$/, '') : x;
const createItem = ({
  id,
  attributes: {
    codeBlocks: [code, example],
    tags,
    text,
  },
}) => mapValues({ code, example, id, tags, text }, enforceSingleNewLine);

(async () => {
  await promisify(fs.writeFile)(
    path.resolve(__dirname, FILE_NAME),
    JSON.stringify(snippets.map(createItem))
  );

  // eslint-disable-next-line no-console
  return console.log(`Successfully created ${FILE_NAME} file.`);
})();
