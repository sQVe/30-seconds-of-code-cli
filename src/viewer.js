const TerminalRenderer = require('marked-terminal');
const chalk = require('chalk');
const marked = require('marked');
const { highlight } = require('cli-highlight');

const { enforceSingleNewLine } = require('./helpers');

marked.setOptions({ renderer: new TerminalRenderer() });

const prettyPrint = layout => {
  const layoutMap = {
    i: x =>
      [
        chalk.magenta.bold(x.id),
        chalk.magenta.bold('-'.repeat(x.id.length)),
      ].join('\n'),
    d: x => enforceSingleNewLine(marked(x.description)),
    c: x => highlight(x.code, { language: 'javascript' }),
    e: x => highlight(x.example, { language: 'javascript' }),
  };

  return x =>
    Array.from(layout)
      .map((k, i) => layoutMap[k](x) + (i > 0 ? '\n' : ''))
      .join('\n');
};

const logSnippet = ({ layout, json }, x) => {
  if (json) {
    console.log(JSON.stringify(x));
  } else {
    console.log(
      (Array.isArray(x) ? x : [x]).map(prettyPrint(layout)).join('\n')
    );
  }
};

module.exports = {
  prettyPrint,
  logSnippet,
};
