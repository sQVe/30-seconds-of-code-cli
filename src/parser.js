#!/usr/bin/env node

//  ┏━┓┏━┓┏━┓┏━┓┏━╸┏━┓
//  ┣━┛┣━┫┣┳┛┗━┓┣╸ ┣┳┛
//  ╹  ╹ ╹╹┗╸┗━┛┗━╸╹┗╸
//
//  Parser for 30 seconds of code snippets.
//  https://github.com/sQVe/30s

import fs from 'fs'
import path from 'path'
import { map } from 'ramda'

import { data as snippets } from '../submodules/30-seconds-of-code/snippet_data/snippets.json'
import { enforceSingleNewLine } from './helpers'

const FILE_NAME = 'snippets.json'

export const createItem = ({
  id,
  attributes: {
    codeBlocks: { es6, example },
    tags,
    text,
  },
}) =>
  map (enforceSingleNewLine, { code: es6, example, id, tags, description: text })
export const writeFile = content =>
  fs.writeFileSync (path.resolve (__dirname, FILE_NAME), JSON.stringify (content))

// eslint-disable-next-line fp/no-unused-expression
writeFile (map (createItem) (snippets))
