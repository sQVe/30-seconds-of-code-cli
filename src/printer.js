import Renderer from 'marked-terminal'
import chalk from 'chalk'
import marked from 'marked'
import { compose, is, join, map, pick, prop, replace, toPairs } from 'ramda'
import { highlight } from 'cli-highlight'
import { writeSync as writeToClipboard } from 'clipboardy'

import { enforceSingleNewLine } from './helpers'

marked.setOptions ({ renderer: new Renderer () })

export const colorizedPrint = x => {
  const printMap = {
    code: y => highlight (y, { language: 'javascript' }),
    description: y => enforceSingleNewLine (marked (y)),
    example: y => highlight (y, { language: 'javascript' }),
    id: y => chalk.magenta.bold (y),
    tags: compose (
      chalk.gray.italic,
      join (', ')
    ),
  }
  const print = y =>
    compose (
      replace (/\n$/, ''),
      join ('\n'),
      map (([k, v]) => printMap[k] (v) + (k === 'id' ? '' : '\n')),
      toPairs
    ) (y)

  return compose (
    join ('\n\n'),
    map (print)
  ) (x)
}

export const printSnippet = ({ cp, layout, json }, x) => {
  const arr = is (Array, x) ? x : [x]
  const layoutMap = {
    c: 'code',
    d: 'description',
    e: 'example',
    i: 'id',
    t: 'tags',
  }
  const keysByLayout = map (k => layoutMap[k]) (Array.from (layout || 'itced'))

  if (x == null) return ''
  if (cp) {
    writeToClipboard (
      compose (
        join ('\n'),
        map (prop ('code'))
      ) (arr)
    )
  }

  const pickedSnippet = map (pick (keysByLayout)) (arr)

  return compose (
    console.log,
    json ? JSON.stringify : colorizedPrint
  ) (pickedSnippet)
}
