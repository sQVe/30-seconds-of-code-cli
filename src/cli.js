#!/usr/bin/env node

//  ┏━┓┏━┓┏━┓
//  ╺━┫┃┃┃┗━┓
//  ┗━┛┗━┛┗━┛
//
//  A CLI for 30 seconds of code snippets.
//  https://github.com/sQVe/30-seconds-of-code-cli

import program from 'commander';

import * as handler from './handler';
import * as viewer from './viewer';
import { version } from '../package.json';

const isTest = process.env.NODE_ENV === 'test';
const actions = {
  tag: (id, opts) => viewer.logSnippet(opts, handler.getSnippet(id)),
  view: (id, opts) => viewer.logSnippet(opts, handler.getSnippetsByTag(id)),
  search: (query, opts) =>
    viewer.logSnippet(opts, handler.searchSnippets(query)),
};
const addCommand = settings => {
  settings.reduce((acc, [key, ...args]) => acc[key](...args), program);
};
const addAction = action => [
  'action',
  isTest
    ? // NOTE: Commander.js sadly does not include a way to hook onto given
      // actions for testing. We solve this by outputting state when NODE_ENV
      // is test.
      (input, { layout, json }) =>
        console.log(JSON.stringify([action, input, layout, !!json]))
    : actions[action],
];
const commonOptions = [
  ['option', '-l, --layout <layout>', 'print in specified layout', 'iced'],
  ['option', '-j, --json', 'output in json format', false],
];

program.version(version);

addCommand([
  ['command', 'search [query]'],
  ['description', ['Fuzzy search snippets']],
  ...commonOptions,
  addAction('search'),
]);

addCommand([
  ['command', 'tag [id]'],
  ['description', ['View snippets related to tag']],
  ...commonOptions,
  addAction('tag'),
]);

addCommand([
  ['command', 'view [id]'],
  ['description', ['View snippet']],
  ...commonOptions,
  addAction('view'),
]);

program.on('--help', () =>
  console.log(
    [
      '',
      'Examples:',
      '',
      '  view all',
      '  search flatten --output json all',
      '  tag --print ic array',
    ].join('\n')
  )
);
program.parse(process.argv);
