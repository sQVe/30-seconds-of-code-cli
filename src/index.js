//  ┏━┓┏━┓┏━┓
//  ╺━┫┃┃┃┗━┓
//  ┗━┛┗━┛┗━┛
//
//  A CLI for 30 seconds of code snippets.
//  https://github.com/sQVe/30s

import program from 'commander'
import { reduce } from 'ramda'

import snippets from '../lib/snippets.json'
import { getSnippet, getSnippetsByTag, searchSnippets } from './handler'
import { printSnippet } from './printer'
import { version } from '../package.json'

const isTest = process.env.NODE_ENV === 'test'
const actions = {
  tag: (id, opts) => printSnippet (opts, getSnippetsByTag (snippets, id)),
  view: (id, opts) => printSnippet (opts, getSnippet (snippets, id)),
  search: (query, opts) => printSnippet (opts, searchSnippets (snippets, query)),
}
const addCommand = settings =>
  reduce ((acc, [key, ...args]) => acc[key] (...args), program, settings)

const addAction = action => [
  'action',
  (input, opts) => {
    if (!input) return program.outputHelp ()

    return isTest
      ? // NOTE: Commander.js sadly does not include a way to hook onto given
        // actions for integration testing. We solve this by outputting state
        // when NODE_ENV is test.
        console.log (
          JSON.stringify ([action, input, !!opts.cp, !!opts.json, opts.layout])
        )
      : actions[action] (input, opts)
  },
]
const commonOptions = [
  ['option', '-c, --cp', 'copy code to clipboard', false],
  ['option', '-j, --json', 'output in json format', false],
  ['option', '-l, --layout <layout>', 'print in specified layout', 'itced'],
]

program.version (version)

addCommand ([
  ['command', 's [query]'],
  ['alias', 'search'],
  ['description', ['fuzzy search snippets by id']],
  ...commonOptions,
  addAction ('search'),
])

addCommand ([
  ['command', 't [id]'],
  ['alias', 'tag'],
  ['description', ['view snippets by tag']],
  ...commonOptions,
  addAction ('tag'),
])

addCommand ([
  ['command', 'v [id]'],
  ['alias', 'view'],
  ['description', ['view snippet with id']],
  ...commonOptions,
  addAction ('view'),
])

program.on ('--help', () =>
  console.log (
    [
      '',
      'Examples:',
      '  v head',
      '  view head',
      '',
      '  s -j flatten',
      '  search --json flatten',
      '',
      '  t -l ic array',
      '  tag --layout ic array',
    ].join ('\n')
  )
)

program.parse (process.argv)
if (process.argv.length <= 2) {
  program.outputHelp ()
}
