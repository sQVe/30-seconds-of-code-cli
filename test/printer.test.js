jest.mock('clipboardy')

import { head } from 'ramda'
import { writeSync as writeToClipboard } from 'clipboardy'

import { printSnippet, colorizedPrint } from '../src/printer'

const layouts = [...Array.from('itced'), 'itced']
const snippet = {
  code: 'const head = arr => arr[0];',
  description: [
    'Returns the head of a list.',
    'Use `arr[0]` to return the first element of the passed array.',
  ].join('\n'),
  example: 'head([1, 2, 3]); // 1',
  id: 'head',
  tags: ['array', 'beginner'],
}

describe('Printer', () => {
  describe('colorizedPrint', () => {
    it('should print in color based on given snippet', () => {
      expect(
        colorizedPrint(
          Object.keys(snippet).map((k) => ({ [k]: snippet[k] }))
        ).toString()
      ).toMatchSnapshot()
      expect(colorizedPrint([snippet])).toMatchSnapshot()
    })
  })

  describe('printSnippet', () => {
    jest.spyOn(global.console, 'log').mockImplementation(() => {
      // Noop.
    })

    it('should handle snippets in a array', () => {
      printSnippet({ layout: 'ic' }, [snippet])

      expect(head(console.log.mock.calls)).toMatchSnapshot()
    })

    it('should handle snippet in object', () => {
      printSnippet({ layout: 'ic' }, snippet)

      expect(head(console.log.mock.calls)).toMatchSnapshot()
    })

    it('should handle no snippet', () => {
      printSnippet({ layout: 'ic' }, undefined)

      expect(head(console.log.mock.calls)).toBeUndefined()
    })

    it('should copy write blocks to clipboard', () => {
      printSnippet({ layout: 'iced', cp: true }, snippet)

      expect(writeToClipboard).toHaveBeenCalledTimes(1)
      expect(writeToClipboard).toHaveBeenCalledWith(snippet.code)
    })

    it('should copy multiple code blocks to clipboard', () => {
      printSnippet({ layout: 'iced', cp: true }, [snippet, snippet])

      expect(writeToClipboard).toHaveBeenCalledTimes(1)
      expect(writeToClipboard).toHaveBeenCalledWith(
        [snippet.code, snippet.code].join('\n')
      )
    })

    layouts.forEach((layout) => {
      describe(`with layout set to "${layout}"`, () => {
        it('should log JSON given json flag', () => {
          printSnippet({ layout, json: true }, snippet)

          expect(head(console.log.mock.calls)).toMatchSnapshot()
        })

        it('should print in color by default', () => {
          printSnippet({ layout }, snippet)

          expect(head(console.log.mock.calls)).toMatchSnapshot()
        })
      })
    })
  })
})
