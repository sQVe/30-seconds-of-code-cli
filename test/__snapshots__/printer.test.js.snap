// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Printer colorizedPrint should print in color based on given snippet 1`] = `
"[34mconst[39m head = [33marr =>[39m arr[[32m0[39m];

[0mReturns the head of a list.[0m
[0mUse [33marr[0][39m to return the first element of the passed array.[0m

head([[32m1[39m, [32m2[39m, [32m3[39m]); [32m// 1[39m

[35m[1mhead[22m[39m

[90m[3marray, beginner[23m[39m"
`;

exports[`Printer colorizedPrint should print in color based on given snippet 2`] = `
"[34mconst[39m head = [33marr =>[39m arr[[32m0[39m];

[0mReturns the head of a list.[0m
[0mUse [33marr[0][39m to return the first element of the passed array.[0m

head([[32m1[39m, [32m2[39m, [32m3[39m]); [32m// 1[39m

[35m[1mhead[22m[39m
[90m[3marray, beginner[23m[39m"
`;

exports[`Printer printSnippet should handle snippet in object 1`] = `
Array [
  "[35m[1mhead[22m[39m
[34mconst[39m head = [33marr =>[39m arr[[32m0[39m];",
]
`;

exports[`Printer printSnippet should handle snippets in a array 1`] = `
Array [
  "[35m[1mhead[22m[39m
[34mconst[39m head = [33marr =>[39m arr[[32m0[39m];",
]
`;

exports[`Printer printSnippet with layout set to "c" should log JSON given json flag 1`] = `
Array [
  "[{\\"code\\":\\"const head = arr => arr[0];\\"}]",
]
`;

exports[`Printer printSnippet with layout set to "c" should print in color by default 1`] = `
Array [
  "[34mconst[39m head = [33marr =>[39m arr[[32m0[39m];",
]
`;

exports[`Printer printSnippet with layout set to "d" should log JSON given json flag 1`] = `
Array [
  "[{\\"description\\":\\"Returns the head of a list.\\\\nUse \`arr[0]\` to return the first element of the passed array.\\"}]",
]
`;

exports[`Printer printSnippet with layout set to "d" should print in color by default 1`] = `
Array [
  "[0mReturns the head of a list.[0m
[0mUse [33marr[0][39m to return the first element of the passed array.[0m",
]
`;

exports[`Printer printSnippet with layout set to "e" should log JSON given json flag 1`] = `
Array [
  "[{\\"example\\":\\"head([1, 2, 3]); // 1\\"}]",
]
`;

exports[`Printer printSnippet with layout set to "e" should print in color by default 1`] = `
Array [
  "head([[32m1[39m, [32m2[39m, [32m3[39m]); [32m// 1[39m",
]
`;

exports[`Printer printSnippet with layout set to "i" should log JSON given json flag 1`] = `
Array [
  "[{\\"id\\":\\"head\\"}]",
]
`;

exports[`Printer printSnippet with layout set to "i" should print in color by default 1`] = `
Array [
  "[35m[1mhead[22m[39m",
]
`;

exports[`Printer printSnippet with layout set to "itced" should log JSON given json flag 1`] = `
Array [
  "[{\\"id\\":\\"head\\",\\"tags\\":[\\"array\\",\\"beginner\\"],\\"code\\":\\"const head = arr => arr[0];\\",\\"example\\":\\"head([1, 2, 3]); // 1\\",\\"description\\":\\"Returns the head of a list.\\\\nUse \`arr[0]\` to return the first element of the passed array.\\"}]",
]
`;

exports[`Printer printSnippet with layout set to "itced" should print in color by default 1`] = `
Array [
  "[35m[1mhead[22m[39m
[90m[3marray, beginner[23m[39m

[34mconst[39m head = [33marr =>[39m arr[[32m0[39m];

head([[32m1[39m, [32m2[39m, [32m3[39m]); [32m// 1[39m

[0mReturns the head of a list.[0m
[0mUse [33marr[0][39m to return the first element of the passed array.[0m",
]
`;

exports[`Printer printSnippet with layout set to "t" should log JSON given json flag 1`] = `
Array [
  "[{\\"tags\\":[\\"array\\",\\"beginner\\"]}]",
]
`;

exports[`Printer printSnippet with layout set to "t" should print in color by default 1`] = `
Array [
  "[90m[3marray, beginner[23m[39m",
]
`;
